samples/ChangeFeedProcessor/DocumentDB.ChangeFeedProcessor/ChangeFeedEventHost.cs

```diff
@@ -297,133 +297,144 @@ public ChangeFeedEventHost(string hostName, DocumentCollectionInfo documentColle
+                                {
+                                    lease = await CheckpointAsync(lease, response.ResponseContinuation, context);
+                                    checkpointStats.Reset();
+                                }
+                                else if (response.Count > 0)
+                                {
+                                    TraceLog.Informational(string.Format("Checkpoint: not checkpointing for partition {0}, {1} docs, new continuation '{2}' as frequency condition is not met", lease.PartitionId, response.Count, response.ResponseContinuation));
+                                }
                             }
-                        } // Outer while (this.isShutdown == 0) loop.
+                        }
+                        while (query.HasMoreResults && this.isShutdown == 0);
 
-                        closeReason = ChangeFeedObserverCloseReason.Shutdown;
-                    }
-                    catch (TaskCanceledException)
-                    {
-                        Debug.Assert(cancellation.IsCancellationRequested, "cancellation.IsCancellationRequested");
-                        TraceLog.Informational(string.Format("Cancel signal received for partition {0} worker!", context.PartitionKeyRangeId));
-                    }
+                        if (this.isShutdown == 0)
+                        {
+                            await Task.Delay(this.options.FeedPollDelay, cancellation.Token);
+                        }
+                    } // Outer while (this.isShutdown == 0) loop.
+
+                    closeReason = ChangeFeedObserverCloseReason.Shutdown;
                 }
                 catch (LeaseLostException ex)
                 {
                     closeReason = ex.IsGone ? ChangeFeedObserverCloseReason.LeaseGone : ChangeFeedObserverCloseReason.LeaseLost;
                 }
+                catch (TaskCanceledException ex)
+                {
+                    if (cancellation.IsCancellationRequested || this.isShutdown != 0)
+                    {
+                        TraceLog.Informational(string.Format("Cancel signal received for partition {0} worker!", context.PartitionKeyRangeId));
+                        if (!closeReason.HasValue)
+                        {
+                            closeReason = ChangeFeedObserverCloseReason.Shutdown;
+                        }
+                    }
+                    else
+                    {
+                        TraceLog.Warning(string.Format("Partition {0}: got task cancelled exception in non-shutdown scenario [cancellation={1}, isShutdown={2}], {3}", context.PartitionKeyRangeId, cancellation.IsCancellationRequested, this.isShutdown, ex.StackTrace));
+                        if (!closeReason.HasValue)
+                        {
+                            closeReason = ChangeFeedObserverCloseReason.Unknown;
+                        }
+                    }
+                }
```

Comment:
Do we need "if (closeReason.HasValue)" at line 477? IMHO, we exit try/catch always with reason, no?

Reply:
That's for safety, doesn't hurt. The idea is that we don't overwrite if it was already set.